### ISSUE TITLE
Math::BigRat in perl-5.26.0-RC2 severely broken

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15991]
Migrated from [rt.perl.org#131410](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=131410) (status was 'open')

Searchable as RT131410$

**Comment by p5pRT** [comment]
### From mhasch-p5p@cozap.com

This is a bug report for perl from mhasch\-p5p@&#8203;cozap\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.26\.0\.




---
Dear Perl 5 Porters\,

The release candidate 2 for perl\-5\.26\.0 comes with Math&#8203;::BigRat
v0\.2611 which is severely broken\.  Neither this version of
Math&#8203;::BigRat nor its current CPAN version v0\.2612 should be
distributed with perl core\.

The impact would be that code using Math&#8203;::BigRat for symbolic
math is likely to break\, as Math&#8203;::BigRat objects will randomly
change their values dependent on calculations with other objects\.
Paul Haffenden has first reported this in CPAN RT \#121139\.
https://rt.cpan.org/Ticket/Display.html?id=121139

Here is a code snippet demonstrating the problem&#8203;:

&nbsp;    my $a = Math&#8203;::BigRat\->new\('3/2'\);
&nbsp;    my $x = Math&#8203;::BigRat\->new\('2/3'\);
&nbsp;    my $y = $a;
&nbsp;    $y = $x \* $y;
&nbsp;    print "BONK\\n" if "$a" ne '3/2';

Note that Math&#8203;::BigRat has always been supporting copying
behaviour for assignments rather than simple reference assignment
so far\.  This is normally achieved by correctly implementing
'=' with the overload pragma\.  Application code relying on this
behaviour is bound to break with the new bug\.

It can be worked around by replacing many assignments like
''$foo = $bar'' by ''$foo = $bar\->copy''\, but this in turn
breaks the level of abstraction provided by overloaded math
operators in the first place\.

As the Math&#8203;::BigXXX family of modules is currently under active
development I expect we will soon have a fix for this problem\.
In case of an emergency I'd rather have a Math&#8203;::BigRat version
without the latest bells and whistles in the Perl 5 core
distribution than something massively breaking existing code\.
A version of Math&#8203;::BigRat known to work would be v0\.260804\,
as in the development track from perl\-5\.25\.4 to perl\-5\.25\.7\.

It is somewhat important not to use an even earlier version\,
since v0\.260804 was the first one to cope with incompatible
changes in Math&#8203;::BigInt released at the same time\.  As the
modules of this family are being developed in parallel but
come in separate CPAN distributions\, there is always a risk
to end up with incompatible versions after updates\.  Perl
core providing fairly recent ones will help to avoid older
iterations of that trap\.  \(See also the discussion in
http&#8203;://blogs\.perl\.org/users/martin/2016/08/the\-math\-bigrat\-trap\.html
\.\)

Thank you for your trouble\,
\-Martin



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=critical
&nbsp;    module=Math&#8203;::BigRat


---
Site configuration information for perl 5\.26\.0&#8203;:

Configured by martin at Wed May 31 12&#8203;:56&#8203;:17 CEST 2017\.

Summary of my perl5 \(revision 5 version 26 subversion 0\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux
&nbsp;    osvers=3\.2\.0\-4\-amd64
&nbsp;    archname=x86\_64\-linux\-thread\-multi\-ld
&nbsp;    uname='linux dewey 3\.2\.0\-4\-amd64 \#1 smp debian 3\.2\.86\-1 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusedevel \-Uversiononly \-Dprefix=/opt/perl526t \-Dusethreads \-Duse64bitint \-Duse64bitall \-Duselongdouble \-Dperladmin=none \-Dusevfork=false \-de'
&nbsp;    hint=recommended
&nbsp;    useposix=true
&nbsp;    d\_sigaction=define
&nbsp;    useithreads=define
&nbsp;    usemultiplicity=define
&nbsp;    use64bitint=define
&nbsp;    use64bitall=define
&nbsp;    uselongdouble=define
&nbsp;    usemymalloc=n
&nbsp;    default\_inc\_excludes\_dot=define
&nbsp;    bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'
&nbsp;    ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-D\_FORTIFY\_SOURCE=2'
&nbsp;    optimize='\-O2'
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''
&nbsp;    gccversion='4\.7\.2'
&nbsp;    gccosandvers=''
&nbsp;    intsize=4
&nbsp;    longsize=8
&nbsp;    ptrsize=8
&nbsp;    doublesize=8
&nbsp;    byteorder=12345678
&nbsp;    doublekind=3
&nbsp;    d\_longlong=define
&nbsp;    longlongsize=8
&nbsp;    d\_longdbl=define
&nbsp;    longdblsize=16
&nbsp;    longdblkind=3
&nbsp;    ivtype='long'
&nbsp;    ivsize=8
&nbsp;    nvtype='long double'
&nbsp;    nvsize=16
&nbsp;    Off\_t='off\_t'
&nbsp;    lseeksize=8
&nbsp;    alignbytes=16
&nbsp;    prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'
&nbsp;    ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/4\.7/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /lib64 /usr/lib64
&nbsp;    libs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.13\.so
&nbsp;    so=so
&nbsp;    useshrplib=false
&nbsp;    libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.13'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs
&nbsp;    dlext=so
&nbsp;    d\_dlsymun=undef
&nbsp;    ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'
&nbsp;    lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    RC2



---
@&#8203;INC for perl 5\.26\.0&#8203;:
&nbsp;    /opt/perl526t/lib/site\_perl/5\.26\.0/x86\_64\-linux\-thread\-multi\-ld
&nbsp;    /opt/perl526t/lib/site\_perl/5\.26\.0
&nbsp;    /opt/perl526t/lib/5\.26\.0/x86\_64\-linux\-thread\-multi\-ld
&nbsp;    /opt/perl526t/lib/5\.26\.0



---
Environment for perl 5\.26\.0&#8203;:
&nbsp;    HOME=/home/martin
&nbsp;    LANG=C
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=en\_US
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/opt/perl526t/bin&#8203;:/opt/perl524/bin&#8203;:/home/martin/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/local/games&#8203;:/usr/games
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, May 31\, 2017 at 07&#8203;:32&#8203;:07AM \-0700\, Martin Hasch wrote&#8203;:
> \# New Ticket Created by  Martin Hasch 
> \# Please include the string&#8203;:  \[perl \#131410\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=131410 >
> 
> 
> This is a bug report for perl from mhasch\-p5p@&#8203;cozap\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.26\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Dear Perl 5 Porters\,
> 
> The release candidate 2 for perl\-5\.26\.0 comes with Math&#8203;::BigRat
> v0\.2611 which is severely broken\.  Neither this version of
> Math&#8203;::BigRat nor its current CPAN version v0\.2612 should be
> distributed with perl core\.

Note that perl 5\.26\.0 was released yesterday evening

\-\- 
You live and learn \(although usually you just live\)\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From mhasch-p5p@cozap.com

On Thu\, Jun 01\, 2017 at 03&#8203;:40&#8203;:21AM \-0700\, Dave Mitchell via RT wrote&#8203;:
> On Wed\, May 31\, 2017 at 07&#8203;:32&#8203;:07AM \-0700\, Martin Hasch wrote&#8203;:
> > The release candidate 2 for perl\-5\.26\.0 comes with Math&#8203;::BigRat
> > v0\.2611 which is severely broken\.  Neither this version of
> > Math&#8203;::BigRat nor its current CPAN version v0\.2612 should be
> > distributed with perl core\.
> 
> Note that perl 5\.26\.0 was released yesterday evening

Yes\, I realized that\, it's a pity I didn't find out earlier\.

The good news is that there is already a patch\.  Peter shouldn't
have too much trouble now to fix the issue and thus make it
into perl\-5\.26\.1\.

\-Martin

**Comment by p5pRT** [comment]
### From @pjacklam

Ouch\, this is bad\. I'll give this top priority\.

As a side remark\, I would like to have Math&#8203;::BigRat in the same distribution as Math&#8203;::BigInt and Math&#8203;::BigFloat as these three modules are so closely related\.

Also\, the fact that Math&#8203;::BigRat isa Math&#8203;::BigFloat\, and Math&#8203;::BigFloat isa Math&#8203;::BigInt is causing never ending problems\. Introducing Math&#8203;::BigReal and letting Math&#8203;::Big\(Int|Float|Rat\) be a subclass of Math&#8203;::BigReal would make maintenance much easier\.

**Comment by p5pRT** [comment]
### From mhasch-p5p@cozap.com

Peter\,

On Fri\, Jun 09\, 2017 at 02&#8203;:11&#8203;:38AM \-0700\, \(Peter J\. Acklam\) via RT wrote&#8203;:
> As a side remark\, I would like to have Math&#8203;::BigRat in the same
> distribution as Math&#8203;::BigInt and Math&#8203;::BigFloat as these three modules
> are so closely related\.

This is a very good idea and will greatly help to avoid problems
like the one reported in rt\.cpan\.org \#122018\, say\.

> Also\, the fact that Math&#8203;::BigRat isa Math&#8203;::BigFloat\, and Math&#8203;::BigFloat
> isa Math&#8203;::BigInt is causing never ending problems\. Introducing
> Math&#8203;::BigReal and letting Math&#8203;::Big\(Int|Float|Rat\) be a subclass of
> Math&#8203;::BigReal would make maintenance much easier\.

I would be delighted if you could make this work\.  Such a new
architecture might also make it easier to take advantage of
backends that implement more multiprecision functionality than
just integers and floating point numbers\.

Let us not get distracted from quickly fixing the problem at
hand though\.  I think I already nailed it with the patch in
rt\.cpan\.org \#121139\, but I am not sure that it is complete\.

\-Martin

**Comment by khwilliamson** [comment]
what is the status of this ticket?

**Comment by jkeenan** [comment]
> what is the status of this ticket?

I'm drowning in BBC tickets this morning, but a quick look at http://fast2-matrix.cpantesters.org/?dist=Math-BigInt (the distribution includes Math::BigRat) suggests that everything is PASSing.  In any case, if anything needs to be done, this is CPAN-upstream.  So I think this ticket is closeable.

**Comment by jkeenan** [comment]
> > what is the status of this ticket?
> 
> I'm drowning in BBC tickets this morning, but a quick look at http://fast2-matrix.cpantesters.org/?dist=Math-BigInt (the distribution includes Math::BigRat) suggests that everything is PASSing. In any case, if anything needs to be done, this is CPAN-upstream. So I think this ticket is closeable.

http://fast2-matrix.cpantesters.org/?dist=Math-BigInt all green.  So, closing.

**Comment by pjacklam** [comment]
I'm sorry I didn't follow up on this. These distributions are maintained upstream, and I tend to miss tickets that aren't reported there.
